File: HW 2/CSE216HW2/TwoDPoint.java
// Nitin Dev
// NDEV
// 112298641

import java.util.ArrayList;
import java.util.List;

/**
 * An unmodifiable point in the standard two-dimensional Euclidean space. The
 * coordinates of such a point is given by exactly two doubles specifying its
 * <code>x</code> and <code>y</code> values.
 */
public class TwoDPoint implements Point {

    private double x;
    private double y;

    public TwoDPoint(double x, double y) {
        // TODO
        this.x = x;
        this.y = y;
    }

    /**
     * @return the coordinates of this point as a <code>double[]</code>.
     */
    @Override
    public double[] coordinates() {
        // TODO
        return new double[] { x, y };
    }

    /**
     * Returns a list of <code>TwoDPoint</code>s based on the specified array of
     * doubles. A valid argument must always be an even number of doubles so that
     * every pair can be used to form a single <code>TwoDPoint</code> to be added to
     * the returned list of points.
     *
     * @param coordinates the specified array of doubles.
     * @return a list of two-dimensional point objects.
     * @throws IllegalArgumentException if the input array has an odd number of
     *                                  doubles.
     */
    public static List<TwoDPoint> ofDoubles(double... coordinates) throws IllegalArgumentException {
        // TODO
        if (coordinates.length % 2 == 1) {
            throw new IllegalArgumentException();
        }
        ArrayList<TwoDPoint> twoDPoints = new ArrayList<>();
        for (int x = 0; x < coordinates.length - 1; x += 2) {
            TwoDPoint temp = new TwoDPoint(coordinates[x], coordinates[x + 1]);
            twoDPoints.add(temp);
        }
        return twoDPoints;
    }

    public double getX() {
        return this.x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return this.y;
    }

    public void setY(double y) {
        this.y = y;
    }
}

File: HW 5/CSE216HW5/two_d_point.py
# Nitin Dev
# NDEV
# 112298641

from typing import List

class TwoDPoint:

    def __init__(self, x, y) -> None:
        self.__x = x
        self.__y = y

    @property
    def x(self) -> float:
        return self.__x

    @property
    def y(self) -> float:
        return self.__y

    def __eq__(self, other: object) -> bool:
        try:
            return (self.__x, self.__y) == (other.x, other.y)
        except:
            return False

    def __ne__(self, other: object) -> bool:
        return not self.__eq__(other)

    def __str__(self) -> str:
        return '(%g, %g)' % (self.__x, self.__y)

    # TODO: add magic methods such that two TwoDPoint objects can be added and subtracted coordinate-wise just by using
    #  syntax of the form p + q or p - q

    def __add__(self, other: object) -> 'TwoDPoint':
        try:
            return TwoDPoint(self.__x + other.x, self.__y + other.y)
        except:
            return None

    def __sub__(self, other: object) -> 'TwoDPoint':
        try:
            return TwoDPoint(self.__x - other.x, self.__y - other.y)
        except:
            return None

    @staticmethod
    def from_coordinates(coordinates: List[float]) -> List['TwoDPoint']:
        if len(coordinates) % 2 != 0:
            raise Exception("Odd number of floats given to build a list of 2-d points")
        points = []
        it = iter(coordinates)
        for x in it:
            points.append(TwoDPoint(x, next(it)))
        return points
